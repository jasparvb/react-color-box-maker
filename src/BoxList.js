import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import "./BoxList.css";


function BoxList() {
    const [boxes, setBoxes] = useState([]);

    /** Add new box */
    const addBox = box => {
        setBoxes(boxes => [...boxes, box]);
    };
    
    
    /** Remove box */
    const removeBox = id => {
        setBoxes(boxes => boxes.filter(box => box.id !== id));
    };

    const allBoxes = boxes.map(b =>
        <Box 
        key={b.id} 
        id={b.id} 
        width={b.width} 
        height={b.height} 
        backgroundColor={b.backgroundColor} 
        removeBox={removeBox}
        />
    );
    
    return (
        <div className="BoxList">
            <NewBoxForm addBox={addBox} />
            <div className="BoxList-Container">{allBoxes}</div>
        </div>
    );
};

export default BoxList;


