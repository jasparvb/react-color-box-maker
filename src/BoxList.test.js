import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

// smoke test
it("renders without crashing", function() {
  render(<BoxList />);
});
// snapshot test
it("matches snapshot", function() {
  const {asFragment} = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});


it("can add a new box", function() {
  const { getByLabelText, queryByText, getByText } = render(<BoxList />);

  // no items yet
  expect(queryByText("X")).not.toBeInTheDocument();

  const widthInput = getByLabelText("Width:");
  const heightInput = getByLabelText("Height:");
  const colorInput = getByLabelText("Background Color:");
  const submitBtn = queryByText("Add a new box!")

  // fill out the form
  fireEvent.change(widthInput, { target: { value: "200px" }});
  fireEvent.change(heightInput, { target: { value: "100px" }});
  fireEvent.change(colorInput, { target: { value: "#b13e3e" }});
  fireEvent.click(submitBtn);

  // item exists!
  const removeBtn = getByText("X");
  expect(removeBtn).toBeInTheDocument();
  expect(removeBtn.previousSibling).toHaveStyle(`
    width: 200px;
    height: 100px;
    background-color: #b13e3e;
  `);

});

it("can remove a new box", function() {
  const { getByLabelText, queryByText, getByText } = render(<BoxList />);

  const widthInput = getByLabelText("Width:");
  const heightInput = getByLabelText("Height:");
  const colorInput = getByLabelText("Background Color:");
  const submitBtn = queryByText("Add a new box!")

  // fill out the form
  fireEvent.change(widthInput, { target: { value: "200px" }});
  fireEvent.change(heightInput, { target: { value: "100px" }});
  fireEvent.change(colorInput, { target: { value: "#b13e3e" }});
  fireEvent.click(submitBtn);
  
  // item exists!
  const removeBtn = getByText("X");
  fireEvent.click(removeBtn);
  expect(removeBtn).not.toBeInTheDocument();
});