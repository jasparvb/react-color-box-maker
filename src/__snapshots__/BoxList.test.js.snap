// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form>
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        value=""
      />
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        value=""
      />
      <label
        for="backgroundColor"
      >
        Background Color:
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        type="color"
        value=""
      />
      <button>
        Add a new box!
      </button>
    </form>
    <div
      class="BoxList-Container"
    />
  </div>
</DocumentFragment>
`;
